Source/Math/ConvolveGeometry.h

```diff
@@ -440,16 +446,19 @@ class ConvolveGeometry final
             bool autoPadCur = autoPad[autoPad.size() == 1 ? 0 : i];
             size_t lo = lowerPad[lowerPad.size() == 1 ? 0 : i];
             size_t hi = upperPad[upperPad.size() == 1 ? 0 : i];
+            size_t dil = dilation[dilation.GetRank() == 1 ? 0 : i];
             if (autoPadCur)
             {
-                dim += kernelShape[i] - 1;
+                dim += dil * (kernelShape[i] - 1);
             }
             else
             {
                 dim += lo + hi;
             }
-            float preciseDimOut = (float)(dim - kernelShape[i]) / delta + 1;
+            float preciseDimOut = (float)(dim - (kernelShape[i]-1) * dil - 1) / delta + 1;
```

Comment:
Shouldn't this "(kernelShape[i]-1) * dil - 1" be "(kernelShape[i]-1) * dil + 1"? If kernel is 5 and dil is 2, then you want 9 not 7, right? Also, with this change you don't need something special for dilation=1.

Reply:
I agree with you. But isn't "dim-[(kernelShape[i]-1) * dil + 1]" is the same as "dim-(kernelShape[i]-1) * dil - 1" here? Which is exactly what I did.

Reply:
Sorry, my bad. But with that, you don't need the other variabkle with NoDilation on them, because when dil is 1, it fall back to the right one.

Here my update:

            size_t dil = dilation[dilation.GetRank() == 1 ? 0 : i];
            if (autoPadCur)
            {
                dim += dil * (kernelShape[i] - 1);
            }
            else
            {
                dim += lo + hi;
            }

            size_t effectiveKernelShape = (kernelShape[i] - 1) * dil + 1;
            float preciseDimOut = (float)(dim - effectiveKernelShape) / delta + 1;
            size_t dimOut = static_cast<size_t>(ceilOutDim ? ceil(preciseDimOut) : floor(preciseDimOut));
            // When LowerPad and/or UpperPad are specified (i.e. > 0), we insist that the kernel applications
            // fill the entire space.
            if (!autoPadCur && (lo > 0 || hi > 0))
            {
                size_t size = (dimOut - 1) * delta + kernelShape[i];
                if (size != dim)
                    InvalidArgument("Convolution requires that kernel fills the entire space if auto-padding is disabled.");
            }

            if (mapCount.size() > 1)
                dimOut *= mapCount[i];
            else if (i == inputShape.GetRank() - 1)
                dimOut *= mapCount[0];

            dimsOutput[i] = dimOut;

Reply:
If you combine this piece of code with your approach of set last dimension of dilation to 1, it should work the same as what I did here.
I basically want treat last dimension differently as if there is no dilation(=1).

The problem is that I don't understand 2 problems:
1. why my code generate [1,1,2] from 1, this was definitely not intended.
2. my code should generate [2,2,2] from 2, but because of above code, the last dimension should be handled specially, so the number in last dimension should have no effect anyway.

Reply:
1. For the [1,1,2] because the fromShape had 2 at the end. 
2. I can make it to do that, what about dilation of 3, should it expand to [3,3,3]? WHat if I have [3,2], how should we expand the channel axis.